src/Effectful/Internal/Monad.hs

```diff
@@ -117,36 +127,46 @@ unsafeEff_ m = unsafeEff $ \_ -> m
 -- environment or 'IO').
 type RunIn es m = forall r. Eff es r -> m r
 
--- | Lower 'Eff' operations into 'IO'.
---
--- /Note:/ the 'RunIn' argument has to be called from the same thread as
--- 'unsafeUnliftEff' ('liftBaseWith' / 'withRunInIO'). Attempting otherwise
--- results in a runtime error, because it cannot be done safely in general
--- (*). If you need to spawn threads, have a look at "Effectful.Async" or create
--- a new 'Eff' environment.
---
--- (*) Consider the following:
---
--- @
--- runState 'x' $ do
---   liftBaseWith $ \run -> forkIO $ do
---     threadDelay 1000000
---     run $ ... (1)
---   ... (2)
--- ...
--- @
---
--- If (1) runs after (2) completed, the State Char effect is no longer in scope,
--- but (1) assumes otherwise and things break horribly.
-unsafeUnliftEff :: (RunIn es IO -> IO a) -> Eff es a
-unsafeUnliftEff f = unsafeEff $ \es -> do
+-- | Lower 'Eff' operations into 'IO' ('SeqUnlift').
+seqUnliftEff :: (RunIn es IO -> IO a) -> Eff es a
+seqUnliftEff f = unsafeEff $ \es -> do
   tid0 <- myThreadId
   f $ \m -> do
     tid <- myThreadId
     if tid == tid0
       then unEff m es
-      else error $ "Running Eff operations in a different thread is "
-                ++ "currently not possible with the unlifting function"
+      else error $ "If you want to use the unlifting function to run "
+                ++ "Eff operations in multiple threads, have a look "
+                ++ "at BoundedConcUnlift or UnboundedConcUnlift."
+
+-- | Lower 'Eff' operations into 'IO' ('BoundedConcUnlift')
+boundedConcUnliftEff :: Int -> (RunIn es IO -> IO a) -> Eff es a
+boundedConcUnliftEff threads f = unsafeEff $ \es0 -> do
+  -- Create a copy of the environment as a template for the other threads to
+  -- use. This can't be done from inside the callback as the environment might
+  -- have changed by then.
+  esTemplate <- cloneEnv es0
+  esMapVar <- do
+    tid0 <- myThreadId
+    newMVar $ M.singleton tid0 es0
+  f $ \m -> do
+    es <- modifyMVar esMapVar $ \esMap -> do
+      tid <- myThreadId
+      case tid `M.lookup` esMap of
+        Just es -> pure (esMap, es)
+        Nothing -> case M.size esMap `compare` threads of
+          LT -> do
+            es <- cloneEnv esTemplate
+            pure (M.insert tid es esMap, es)
+          EQ -> pure (M.insert tid esTemplate esMap, esTemplate)
+          GT -> error $ "Number of allowed threads to run with the unlifting "
+                     ++ "function exceeded. You have to increase the argument "
+                     ++ "to BoundedConcUnlift or use UnboundedConcUnlift."
+    unEff m es
+
+-- | Lower 'Eff' operations into 'IO' ('UnboundedConcUnlift').
+unboundedConcUnliftEff :: (RunIn es IO -> IO a) -> Eff es a
+unboundedConcUnliftEff = boundedConcUnliftEff maxBound
```

Comment:
Not sure if it makes that much of a difference, but IMHO it would be better not to define `unboundedConcUnliftEff` in terms of `boundedConcUnliftEff`. Admitted, duplicating the code just to get rid of `M.size`, `compare` and the case statement might be a bit over the top but I think it is worth considering.

Reply:
The cost of comparing sizes is insignificant when compared to cost of cloning the environment, so it doesn't matter.

I'm actually thinking of removing `UnboundedConcUnlift` and leaving just `SeqUnlift` and `ConcUnlift Int` constructors, then setting Int to `maxBound` gives you unbounded unlift.

Reply:
Yeah, this really was a nitpicking comment :-) Regarding the removal of `UnboundedConcUnlift`: I gave it a thought and I think the only reason to keep it is convenience. Note that the current implementation is in fact a bit misleading: Since we never delete `ThreadIds` (should we?) the number of calls in different threads is bound by `maxBound` anyway due to the underlying implementation using a `Data.Map`. But I guess this is mostly a theoretical problem.

Reply:
Dead threads are now getting cleaned up in the unbounded case, so `bounded` and `unbounded` distinction will stay.